Pedestrian randomization feature in random_test_runner

kielczykowski-rai opened the pull request:
## Types of PR

- [x] New Features
- [ ] Upgrade of existing features
- [ ] Bugfix

## Link to the issue
https://tier4.atlassian.net/browse/RJD-740

## Description
This is the implementation of random pedestrians' behavior in random_test_runner.
Pedestrians randomization can be controlled by the following parameters:
* pedestrians count
* planner used for pedestrians control
* min and max speed randomization limits
* enabling static, walking on crosswalks and free walking along vehicle lanes behavior (each can be turned on and off in different configurations
* spawn point offset from lanelets - to make the spawning nearby vehicle lanelets, but make sure that pedestrian is outside the lane

The pedestrians does not cross lanelets. A Jaywalking feature can be an idea to extend current functionality

The current implementation is not fully functional due to SS2 goal setting restrictions - pedestrian goal pose or route points are required to be inside lanelet https://github.com/tier4/scenario_simulator_v2/blob/06c17a06c5a3932a2d5c2ab84da605c1415aca43/simulation/traffic_simulator/src/entity/pedestrian_entity.cpp#L145


## How to review this PR.
[I have temporarily changed the logic](ab23df523955af5ed9c693ba9b1dd849256135ba) to enable goal setting for a single position to acquire. However, a better solution should be implemented to enable route with points outside lanelets setting. - **_It should be decided if such a change comes in with this code or should be implemented in a separate PR._**
## Others

hakuturu583 commented on test_runner/random_test_runner/include/random_test_runner/data_types.hpp:
```diff
@@ -135,19 +170,82 @@ DEFINE_FMT_FORMATTER(
+struct fmt::formatter<NPCPedestrianDescription>
+{
+  template <typename ParseContext>
+  constexpr auto parse(ParseContext & ctx)
```
```suggestion
  constexpr auto parse(const ParseContext & ctx)
```

hakuturu583 commented on test_runner/random_test_runner/include/random_test_runner/data_types.hpp:
```diff
@@ -135,19 +170,82 @@ DEFINE_FMT_FORMATTER(
+struct fmt::formatter<PedestrianBehavior>
+{
+  template <typename ParseContext>
+  constexpr auto parse(ParseContext & ctx)
```
```suggestion
  constexpr auto parse(const ParseContext & ctx)
```

hakuturu583 commented on test_runner/random_test_runner/include/random_test_runner/file_interactions/yaml_test_params_saver.hpp:
```diff
@@ -40,11 +40,20 @@ struct convert<TestSuiteParameters>
-    node["npc_max_speed"] = rhs.npc_max_speed;
-    node["npc_min_spawn_distance_from_ego"] = rhs.npc_min_spawn_distance_from_ego;
-    node["npc_max_spawn_distance_from_ego"] = rhs.npc_max_spawn_distance_from_ego;
+    node["npc_vehicle_count"] = rhs.npc_vehicle_count;
+    node["npc_vehicle_min_speed"] = rhs.npc_vehicle_min_speed;
+    node["npc_vehicle_max_speed"] = rhs.npc_vehicle_max_speed;
+    node["npc_vehicle_min_spawn_distance_from_ego"] = rhs.npc_vehicle_min_spawn_distance_from_ego;
+    node["npc_vehicle_max_spawn_distance_from_ego"] = rhs.npc_vehicle_max_spawn_distance_from_ego;
+    node["npc_pedestrian_count"] = rhs.npc_pedestrian_count;
+    node["npc_pedestrian_planner"] = rhs.npc_pedestrian_planner;
+    node["npc_pedestrian_min_speed"] = rhs.npc_pedestrian_min_speed;
+    node["npc_pedestrian_max_speed"] = rhs.npc_pedestrian_max_speed;
+    node["npc_pedestrian_behavior_static"] = rhs.npc_pedestrian_behavior_static;
+    node["npc_pedestrian_behavior_crosswalk"] = rhs.npc_pedestrian_behavior_crosswalk;
+    node["npc_pedestrian_behavior_walk_along_lane"] = rhs.npc_pedestrian_behavior_walk_along_lane;
+    node["npc_pedestrian_lanelet_min_offset"] = rhs.npc_pedestrian_lanelet_min_offset;
+    node["npc_pedestrian_lanelet_min_offset"] = rhs.npc_pedestrian_lanelet_min_offset;
```
The number of parameters is huge and the documentation and implementation points are far apart, which may cause a large discrepancy between the documentation and the implementation.
Instead of defining parameters directly in yaml, please implement them using [generate_parameter_library](https://github.com/PickNikRobotics/generate_parameter_library), which can automatically generate C++ source code for rosparam from yaml.

hakuturu583 commented on test_runner/random_test_runner/src/lanelet_utils.cpp:
```diff
@@ -166,6 +177,53 @@ struct LaneletPartForRouting
 
+std::vector<geometry_msgs::msg::Pose> LaneletUtils::getPositionsOnNextLanelets(
+  const traffic_simulator_msgs::msg::LaneletPose & start_lanelet_pose,
+  const int64_t & max_number_of_points, double offset)
```
```suggestion
  const int64_t & max_number_of_points, const double offset)
```

hakuturu583 commented on test_runner/random_test_runner/include/random_test_runner/lanelet_utils.hpp:
```diff
@@ -58,11 +59,17 @@ class LaneletUtils
 
+  std::vector<geometry_msgs::msg::Pose> getPositionsOnNextLanelets(
+    const traffic_simulator_msgs::msg::LaneletPose & start_lanelet,
+    const int64_t & max_number_of_points, double offset = 0.0);
```
```suggestion
    const int64_t & max_number_of_points, const double offset = 0.0);
```

hakuturu583 commented on test_runner/random_test_runner/include/random_test_runner/lanelet_utils.hpp:
```diff
@@ -58,11 +59,17 @@ class LaneletUtils
   std::vector<int64_t> getRoute(int64_t from_lanelet_id, int64_t to_lanelet_id);
   double getLaneletLength(int64_t lanelet_id);
   bool isInLanelet(int64_t lanelet_id, double s);
+  std::optional<traffic_simulator_msgs::msg::LaneletPose> toLaneletPose(
+    geometry_msgs::msg::Pose global_pose, bool include_crosswalk = false);
```
A function is implemented that simply calls a member of the HdMapUtils class.
The LaneletUtils::toLaneletPose function is called only once inside LaneletUtils, so there is no need to make it a function.

Even if you must make it a function, avoid unnecessary copying.

kielczykowski-rai replied:
The function was needed in the [test_randomizer.cpp](https://github.com/tier4/scenario_simulator_v2/pull/1151/files#diff-0349381497ffdc3771f21d77972e127ae4056d912db6a2a9e6ce786f85216d29R345). However I will look into the way how to simplify the usage of LaneletUtils and see if it is possible to merge with  traffic_simulator::HdMapUtils

hakuturu583 commented on test_runner/random_test_runner/include/random_test_runner/test_randomizer.hpp:
```diff
@@ -51,10 +52,30 @@ class TestRandomizer
+    const std::vector<traffic_simulator_msgs::msg::LaneletPose> & crosswalk_poses);
+  PedestrianBehavior getRandomPedestrianBehavior(
+    const std::vector<PedestrianBehavior> & pedestrian_behaviors);
+  std::pair<traffic_simulator_msgs::msg::LaneletPose, std::vector<geometry_msgs::msg::Pose>>
+  generatePedestrianSpawnPointAndRouteFromBehavior(
+    const PedestrianBehavior & behavior,
+    const std::vector<traffic_simulator_msgs::msg::LaneletPose> & poses, double min_distance,
+    const std::vector<LaneletPart> & lanelets,
+    const std::vector<traffic_simulator_msgs::msg::LaneletPose> & crosswalk_poses);
```
```suggestion
  auto generatePedestrianSpawnPointAndRouteFromBehavior(
    const PedestrianBehavior & behavior,
    const std::vector<traffic_simulator_msgs::msg::LaneletPose> & poses, const double min_distance,
    const std::vector<LaneletPart> & lanelets,
    const std::vector<traffic_simulator_msgs::msg::LaneletPose> & crosswalk_poses)
    -> std::pair<traffic_simulator_msgs::msg::LaneletPose, std::vector<geometry_msgs::msg::Pose>>;
```

hakuturu583 commented on test_runner/random_test_runner/src/test_randomizer.cpp:
```diff
@@ -157,25 +197,156 @@ traffic_simulator_msgs::msg::LaneletPose TestRandomizer::generateRandomPosition(
+  return pedestrian_behaviors[pedestrian_behavior_id_randomizer.generate()];
+}
+
+std::pair<traffic_simulator_msgs::msg::LaneletPose, std::vector<geometry_msgs::msg::Pose>>
+TestRandomizer::generatePedestrianSpawnPointAndRouteFromBehavior(
+  const PedestrianBehavior & behavior,
+  const std::vector<traffic_simulator_msgs::msg::LaneletPose> & poses, double min_distance,
+  const std::vector<LaneletPart> & lanelets,
+  const std::vector<traffic_simulator_msgs::msg::LaneletPose> & crosswalk_poses)
```
```suggestion
auto TestRandomizer::generatePedestrianSpawnPointAndRouteFromBehavior(
  const PedestrianBehavior & behavior,
  const std::vector<traffic_simulator_msgs::msg::LaneletPose> & poses, const double min_distance,
  const std::vector<LaneletPart> & lanelets,
  const std::vector<traffic_simulator_msgs::msg::LaneletPose> & crosswalk_poses)
  -> std::pair<traffic_simulator_msgs::msg::LaneletPose, std::vector<geometry_msgs::msg::Pose>>
```

hakuturu583 commented on test_runner/random_test_runner/src/test_randomizer.cpp:
```diff
@@ -157,25 +197,156 @@ traffic_simulator_msgs::msg::LaneletPose TestRandomizer::generateRandomPosition(
+}
+
+traffic_simulator_msgs::msg::LaneletPose TestRandomizer::getRandomPedestrianSpawnPose(
+  const std::vector<traffic_simulator_msgs::msg::LaneletPose> & poses, double min_distance,
+  const std::vector<LaneletPart> & lanelets)
```
```suggestion
  const std::vector<traffic_simulator_msgs::msg::LaneletPose> & poses, const double min_distance,
  const std::vector<LaneletPart> & lanelets)
```

hakuturu583 commented on test_runner/random_test_runner/include/random_test_runner/test_randomizer.hpp:
```diff
@@ -51,10 +52,30 @@ class TestRandomizer
+    const std::vector<LaneletPart> & lanelets,
+    const std::vector<traffic_simulator_msgs::msg::LaneletPose> & crosswalk_poses);
+  traffic_simulator_msgs::msg::LaneletPose getRandomPedestrianSpawnPose(
+    const std::vector<traffic_simulator_msgs::msg::LaneletPose> & poses, double min_distance,
```
```suggestion
    const std::vector<traffic_simulator_msgs::msg::LaneletPose> & poses, const double min_distance,
```

hakuturu583 commented on test_runner/random_test_runner/launch/random_test.launch.py:
```diff
@@ -122,13 +116,23 @@ def __init__(self):
-                                        "description":  "Minimum distance of generated npcs from ego"},
-            "npc_max_distance_to_ego": {"default": 100.0,
-                                        "description": "Maximum distance of generated npcs from ego"},
+            "npc_vehicle_count": {"default": 10, "description": "Generated vehicle npc count"},
+            "npc_vehicle_min_speed": {"default": 0.5, "description": "Minimum speed of generated vehicle npcs"},
+            "npc_vehicle_max_speed": {"default": 3.0, "description": "Maximum speed of generated vehicle npcs"},
+            "npc_min_spawn_distance_from_ego": {"default": 10.0,
+                                        "description":  "Minimum distance of generated vehicle npcs from ego"},
+            "npc_vehicle_max_spawn_distance_from_ego": {"default": 110.0,
+                                        "description": "Maximum distance of generated vehicle npcs from ego"},
+            "npc_pedestrian_count": {"default": 10, "description": "Generated pedestrian npc count"},
+            "npc_pedestrian_planner": {"default": "context_gamma_planner", "description": "Pedestrian NPCs planner",
+                               "values": ["context_gamma_planner", "behavior_tree", "default", "do_nothing"]},
+            "npc_pedestrian_min_speed": {"default": 0.1, "description": "Minimum speed of generated pedestrian npcs"},
+            "npc_pedestrian_max_speed": {"default": 1.0, "description": "Maximum speed of generated vehicle npcs"},
+            "npc_pedestrian_behavior_static": {"default": True, "description": "Static pedestrian spawned, no movement applied"},
+            "npc_pedestrian_behavior_crosswalk": {"default": True, "description": "Pedestrian is spawned at crosswalk and passes it"},
+            "npc_pedestrian_behavior_walk_along_lane": {"default": True, "description": "Pedestrian is spawned outside of lanes and moves along them"},
+            "npc_pedestrian_lanelet_min_offset": {"default": 3.0, "description": "Pedestrian minimum spawn point offset from lanelet"},
+            "npc_pedestrian_lanelet_max_offset": {"default": 5.0, "description": "Pedestrian maximum spawn point offset from lanelet"},
```
The number of parameters is increasing and it is becoming difficult to manage them in the launch file.
Please change the specification to read rosparam defined in yaml.

hakuturu583 commented on test_runner/random_test_runner/src/lanelet_utils.cpp:
```diff
@@ -132,7 +137,13 @@ std::optional<traffic_simulator_msgs::msg::LaneletPose> LaneletUtils::getOpposit
+  return toLaneletPose(global_pose, false);
+}
+
+std::optional<traffic_simulator_msgs::msg::LaneletPose> LaneletUtils::toLaneletPose(
+  geometry_msgs::msg::Pose global_pose, bool include_crosswalk)
+{
+  return hdmap_utils_ptr_->toLaneletPose(global_pose, include_crosswalk);
```
```suggestion
```

kielczykowski-rai replied:
[cross reference](https://github.com/tier4/scenario_simulator_v2/pull/1151/files#r1429697267
)

hakuturu583 commented on test_runner/random_test_runner/src/random_test_runner.cpp:
```diff
@@ -113,13 +113,28 @@ TestSuiteParameters RandomTestRunner::collectTestSuiteParameters()
-    this->declare_parameter<double>("npc_min_spawn_distance_from_ego", 10.0);
-  tp.npc_max_spawn_distance_from_ego =
-    this->declare_parameter<double>("npc_max_spawn_distance_from_ego", 100.0);
+  tp.npc_vehicle_count = this->declare_parameter<int>("npc_vehicle_count", 10);
+  tp.npc_vehicle_min_speed = this->declare_parameter<double>("npc_vehicle_min_speed", 0.5);
+  tp.npc_vehicle_max_speed = this->declare_parameter<double>("npc_vehicle_max_speed", 3.0);
+  tp.npc_vehicle_min_spawn_distance_from_ego =
+    this->declare_parameter<double>("npc_vehicle_min_spawn_distance_from_ego", 10.0);
+  tp.npc_vehicle_max_spawn_distance_from_ego =
+    this->declare_parameter<double>("npc_vehicle_max_spawn_distance_from_ego", 100.0);
+  tp.npc_pedestrian_count = this->declare_parameter<int64_t>("npc_pedestrian_count", 1);
+  tp.npc_pedestrian_planner =
+    this->declare_parameter<std::string>("npc_pedestrian_planner", "context_gamma_planner");
+  tp.npc_pedestrian_min_speed = this->declare_parameter<double>("npc_pedestrian_min_speed", 0.1);
+  tp.npc_pedestrian_max_speed = this->declare_parameter<double>("npc_pedestrian_max_speed", 1.0);
+  tp.npc_pedestrian_behavior_static =
+    this->declare_parameter<bool>("npc_pedestrian_behavior_static", true);
+  tp.npc_pedestrian_behavior_crosswalk =
+    this->declare_parameter<bool>("npc_pedestrian_behavior_crosswalk", true);
+  tp.npc_pedestrian_behavior_walk_along_lane =
+    this->declare_parameter<bool>("npc_pedestrian_behavior_walk_along_lane", true);
+  tp.npc_pedestrian_lanelet_min_offset =
+    this->declare_parameter<double>("npc_pedestrian_lanelet_min_offset", 3.0);
+  tp.npc_pedestrian_lanelet_max_offset =
+    this->declare_parameter<double>("npc_pedestrian_lanelet_max_offset", 5.0);
```
Please use [generate_parameters library](https://github.com/PickNikRobotics/generate_parameter_library) and simplify your code.

hakuturu583 reviewed (changes requested):
LaneletUtils has a duplicated role with traffic_simulator::HdMapUtils class and is considered unnecessary.
Please consider integration and deletion.

kielczykowski-rai commented:
@hakuturu583 
Thank you for the review. Many things were caught and corrected thanks to your pertinent comments.

There are still 2 matters that should be more explained on my side:
* `LaneletUtils` class
* extensive parameters in the project

As you pointed out `LaneletUtils` class mostly extends the `traffic_simulator::HdMapUtils` functionality and sometimes only mediates the invocation of functions contained in `traffic_simulator::HdMapUtils`. Such a design of `LaneletUtils` has the following justification:
* by creating a separate class we do not put additional functionality to `traffic_simulator::HdMapUtils`, which already contains many functions and lines of code
* `LaneletUtils` is somewhat an interface between `random_test_runner` and `traffic_simulator`. By gathering all needed functionalities in `LaneletUtils `we expect the functions only to be reused through the usage of this specific class. This creates visible separation of functionalities used only in `random_test_runer` which makes the code and changes easier to track.

However, if the functionalities contained in `LaneletUtils` class are supposed to be reused all over the `traffic_simulator` or `scenario_simulator_v2` we can work on getting rid of the `LaneletUtils` class. I believe such a change would require more workload and can be a good idea for a new PR since extending this PR would become harder to review with every additional change :bow: 

----

I have implemented the parameters using [generate_parameters library](https://github.com/PickNikRobotics/generate_parameter_library). I believe it is a good change and will create more control over errors detection throughout the project. However, there is still a functionality of re-running the tests in `random_test_runner` which requires existence of [YamlTestParamsIO](https://github.com/tier4/scenario_simulator_v2/blob/0ba20f90dd4710672a959a34e25c9615b589250f/test_runner/random_test_runner/include/random_test_runner/file_interactions/yaml_test_params_saver.hpp#L95) class. It requires much more work to redesign the current solution since [generate_parameters library](https://github.com/PickNikRobotics/generate_parameter_library) does not provide any parameters serialization out-of-the-box. :bow: 

Please let me know how should I proceed with moving forward with the PR :bow:

hakuturu583 commented:
I do not think the LaneletUtils class should exist.
The source code in the `random_test_runner` package should be purely scenario related.
It should not contain any operations related to `Lanelet2`.

hakuturu583 commented:
This pull request is no-longer active.
